File: src/api/response.py
"""Research response domain objects."""
from __future__ import annotations

from abc import ABC
from abc import abstractmethod
import re
from typing import Final
from urllib.parse import parse_qsl
from urllib.parse import urlencode
from urllib.parse import urlparse
from urllib.parse import urlunparse

from parallel.types import TaskRunResult
from parallel.types import TaskRunTextOutput

from src.domain.result import Source


class Responded(ABC):
    """Object representing API response."""

    @abstractmethod
    def identifier(self) -> str:
        """Return run identifier."""
        ...

    @abstractmethod
    def cost(self) -> float:
        """Return total request cost."""
        ...

    @abstractmethod
    def raw(self) -> dict:
        """Return raw response payload."""
        ...


class ResearchResponse(Responded):
    """Immutable response from deep research SDK."""

    def __init__(
        self,
        identifier: str,
        status: str,
        output: str,
        basis: list,
        cost: float = 0.0,
        raw: dict = {},
    ) -> None:
        """Initialize with response data."""
        self._identifier: Final[str] = identifier
        self._status: Final[str] = status
        self._output: Final[str] = self._strip(output)
        self._basis: Final[list] = basis
        self._cost: Final[float] = cost
        self._raw: Final[dict] = raw

    def identifier(self) -> str:
        """Return run identifier."""
        return self._identifier

    def cost(self) -> float:
        """Return total request cost."""
        return self._cost

    def raw(self) -> dict:
        """Return raw response payload."""
        return self._raw

    def completed(self) -> bool:
        """Return True if research completed successfully."""
        return self._status == "completed"

    def failed(self) -> bool:
        """Return True if research failed."""
        return self._status == "failed"

    def markdown(self) -> str:
        """Return research output as markdown."""
        return self._output

    def sources(self) -> tuple[Source, ...]:
        """Extract sources from basis citations with confidence."""
        seen: set[str] = set()
        result: list[Source] = []
        for field in self._basis:
            citations = self._value(field, "citations", []) or []
            confidence = self._value(field, "confidence", None)
            for citation in citations:
                link = self._value(citation, "url", "")
                if link:
                    link = self._clean(link)
                if link and link not in seen:
                    seen.add(link)
                    excerpts = self._value(citation, "excerpts", []) or []
                    excerpt = excerpts[0] if excerpts else ""
                    result.append(
                        Source(
                            title=self._value(citation, "title", "") or self._domain(link),
                            url=link,
                            excerpt=excerpt,
                            confidence=confidence,
                        )
                    )
        return tuple(result)

    def _domain(self, url: str) -> str:
        """Extract domain from URL."""
        parsed = urlparse(url)
        return parsed.netloc.replace("www.", "")

    def _strip(self, text: str) -> str:
        """Remove tracking parameters from URLs in text."""
        pattern = r"https?://[^\s\)\]]+"
        result: list[str] = []
        last = 0
        for match in re.finditer(pattern, text):
            result.append(text[last:match.start()])
            result.append(self._clean(match.group(0)))
            last = match.end()
        result.append(text[last:])
        return self._purge("".join(result))

    def _clean(self, url: str) -> str:
        """Remove utm parameters from URL."""
        parts = urlparse(url)
        if not parts.query:
            return url
        items = parse_qsl(parts.query, keep_blank_values=True)
        keep = [(key, value) for key, value in items if not key.lower().startswith("utm_")]
        if len(keep) == len(items):
            return url
        query = urlencode(keep, doseq=True)
        parts = parts._replace(query=query)
        return urlunparse(parts)

    def _purge(self, text: str) -> str:
        """Remove leftover utm fragments from text."""
        return re.sub(r"(\?utm_[^\s\)\]]+|&utm_[^\s\)\]]+)", "", text)

    def _value(self, item, name: str, default):
        """Return attribute or mapping value with fallback."""
        value = getattr(item, name, None)
        if value is not None:
            return value
        try:
            return item[name]
        except Exception:
            return default

    @classmethod
    def parse(cls, result: TaskRunResult) -> ResearchResponse:
        """Create response from SDK result."""
        output = result.output
        content = output.content if isinstance(output, TaskRunTextOutput) else ""
        basis = output.basis if isinstance(output, TaskRunTextOutput) else []
        raw = {}
        try:
            raw = result.model_dump()
        except Exception:
            try:
                raw = result.dict()
            except Exception:
                raw = {}
        return cls(
            identifier=result.run.run_id,
            status=result.run.status,
            output=content,
            basis=basis,
            cost=0.0,
            raw=raw,
        )

File: src/research/api/response.clj
(ns research.api.response
  (:require [clojure.string :as str]
            [research.domain.result :as result]))

(defprotocol Responded
  "Object representing API response."
  (id [item] "Return run identifier.")
  (cost [item] "Return total cost.")
  (raw [item] "Return raw response map.")
  (completed [item] "Return true when completed.")
  (failed [item] "Return true when failed.")
  (text [item] "Return output markdown.")
  (sources [item] "Return source list."))

(defn purge
  "Remove utm fragments from text."
  [text]
  (str/replace text #"(\?utm_[^\s\)\]]+|&utm_[^\s\)\]]+)" ""))

(defn clean
  "Remove utm params from URL."
  [text]
  (try
    (let [part (java.net.URI. text)
          query (.getQuery part)
          value (if (and query (not (str/blank? query)))
                  (let [pairs (map #(str/split % #"=") (str/split query #"&"))
                        items (filter
                               (fn [item]
                                 (not
                                  (str/starts-with?
                                   (str/lower-case (first item))
                                   "utm_")))
                               pairs)
                        line (str/join "&" (map #(str/join "=" %) items))]
                    (if (= line query)
                      text
                      (str (.getScheme part)
                           "://"
                           (.getAuthority part)
                           (.getPath part)
                           (if (str/blank? line) "" (str "?" line))
                           (if (.getFragment part)
                             (str "#" (.getFragment part))
                             ""))))
                  text)]
      value)
    (catch Exception _ text)))

(defn strip
  "Strip tracking URLs from output text."
  [text]
  (let [pattern #"https?://[^\s\)\]]+"
        items (or (re-seq pattern text) [])
        value (reduce
               (fn [note link]
                 (str/replace note link (clean link)))
               text
               items)]
    (purge value)))

(defn domain
  "Extract domain from URL string."
  [text]
  (try (let [part (java.net.URI. text)
             host (.getHost part)]
         (if host (str/replace host "www." "") ""))
       (catch Exception _ "")))

(defrecord Response [id status text data]
  Responded
  (id [_] id)
  (cost [_] (:cost data))
  (raw [_] (:raw data))
  (completed [_] (= status "completed"))
  (failed [_] (= status "failed"))
  (text [_] text)
  (sources [_] (let [seen (atom #{})]
                 (reduce
                  (fn [list field]
                    (let [items (get field :citations [])
                          level (or (get field :confidence "") "")]
                      (reduce
                       (fn [list cite]
                         (let [link (or (:url cite) "")
                               link (if (str/blank? link) "" (clean link))
                               add (and (not (str/blank? link))
                                        (not (contains? @seen link)))]
                           (if add
                             (let [ex (get cite :excerpts [])
                                   note (if (seq ex) (first ex) "")
                                   head (or (:title cite) (domain link))
                                   item (result/->Source
                                         head
                                         link
                                         note
                                         level)]
                               (swap! seen conj link)
                               (conj list item))
                             list)))
                       list
                       items)))
                  []
                  (or (:basis data) [])))))

(defn response
  "Create response from map."
  [item]
  (let [text (strip (or (:output item) ""))
        base {:basis (or (:basis item) [])
              :cost (or (:cost item) 0.0)
              :raw (or (:raw item) {})}]
    (->Response (:id item) (:status item) text base)))
